### DISCUSSION TITLE
Details for capturing USB traffic and PTP message details

### DISCUSSION THREAD

**Discussion by gadgetmies** [frank26080115/alpha-fairy#36] [Q&A]
Hi!

I am trying to build a shoulder display for the Sony A7iii using a STM32 microcontroller and thought I could sniff the PTP packages with Wireshark and replicate those in order to receive notification messages from the camera. However I am unable to capture the USB traffic for some reason although I see and am able to use the XHC20 device. There were some mentions in the [reverse engineering document](https://github.com/frank26080115/alpha-fairy/blob/main/doc/Camera-Reverse-Engineering.md) in this repository of capturing the USB traffic, so I was wondering if someone could point me in the right direction to get started. I'd hugely appreciate also any info on the process that's needed in order to receive the details (aperture, exposure, ISO) from the camera when the settings are changed!

**Comment by frank26080115** [comment]
PTP uses the same data whether you are using USB or IP

The data you are talking about is called "properties", the shutter speed is a "property", and it has a property code `SONYALPHA_PROPCODE_ShutterSpeed       = 0xD20D`

The code constantly requests this and the camera constantly replies with a gigantic packet with a long list of properties. Each data item varies in length, so the parser for this is quite complicated. I call it the property decoder, https://github.com/frank26080115/alpha-fairy/blob/main/arduino_workspace/libraries/PtpIpCamera/PtpIpSonyAlphaCameraPropDecoder.cpp

I have already uploaded the two most important official documentation on PTP and these property packets https://github.com/frank26080115/alpha-fairy/tree/main/ext_docs

I use a Total Phase Beagle USB Analyzer. I have many old capture log files from a A6600 camera. I have uploaded one here. You can open it with the Total Phase Data Center software. https://www.totalphase.com/products/data-center/
[example_a6600_capture.zip](https://github.com/frank26080115/alpha-fairy/files/14132640/example_a6600_capture.zip)

The device property packets I am talking about are the ones that are around 1688 bytes long, they are easy to spot.

This log also includes the whole enumeration sequence.

Have fun.

**Reply by gadgetmies** [reply]
Thank you for sharing the info and sorry that I am able to reply only now. I think I now have a reasonably good idea of how the protocol works. Will however need to figure out the USB side still. I noticed that there’s a ESP32 S3 version of M5Stack Atom. Was wondering if instead of STM32 I should perhaps go with that instead. Seems it also has an IMU and buttons can be added using the GPIO, so perhaps the code in this project could be used as basis for the implementation. Do you think it would make sense to add support for the ATOMS3 and USB control to this project? That might even improve the device support.

**Reply by frank26080115** [reply]
I have no experience with the ESP32-S3's USB host library. Also, I am suspicious of the ATOMS3. Most of the dev boards around the S3 chip make it not possible to use the USB host functionality in order to make debugging easier. You need to make sure it can actually do USB host.

I also think the S3 won't work because it has no battery. You'll need to add a battery, and you'll also need to make sure the camera doesn't start draining your battery because it starts to recharge itself through the USB port. I think there's a menu option on the camera to disable that but it's important to remember.

I have definitely already implemented camera controls for my A6600 on a Teensy 3, using https://github.com/PaulStoffregen/USBHost_t36 , it was pretty straight forward to accomplish. I did use a gigantic USB power bank for that project.

I also know STM32's USB host library is very powerful and flexible, I have done huge projects with it before (hosting multiple devices, through a hub)
